### PR TITLE
Decouple resource provider registration

### PR DESCRIPTION
**Pull request by kesselb**
https://github.com/nextcloud/server/blob/a0d48c568d34d20a894b55ef2d26d972339866a2/apps/files/lib/AppInfo/Application.php#L88-L89

Apps can register their own `ResourceProvider`. For now they get a `OC\Collaboration\Resources\Manager` instance with `IDBConnection` as dependency. But a database connection is not required for `registerResourceProvider`. This pr extract the register and provider resolver functionality. I don't know if there are some naming conventions. Feel free to suggest better names ;)

Discovered this while debugging https://github.com/nextcloud/server/pull/13712.

### CODE REVIEW

File: lib/public/Collaboration/Resources/IProviderManager.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace OCP\Collaboration\Resources;
+
+/**
+ * @since 18.0.0
+ */
+interface IProviderManager {
+
+	/**
+	 * @return array list of resource providers
```
Comment (ChristophWurst) [inline lib/public/Collaboration/Resources/IProviderManager.php]:
```suggestion
	 * @return IProvider[] list of resource providers
```

File: lib/public/Collaboration/Resources/IProviderManager.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	 * @since 18.0.0
+	 */
+	public function getResourceProviders(): array;
+
+	/**
+	 * @param string $provider
```
Comment (ChristophWurst) [inline lib/public/Collaboration/Resources/IProviderManager.php]:
```suggestion
	 * @param string $provider provider's class name
```

File: lib/public/Collaboration/Resources/IManager.php [hunk 3]
```diff
@@ -116,6 +116,7 @@ public function getResourceForUser(string $type, string $id, ?IUser $user): IRes
 	/**
 	 * @param string $provider
 	 * @since 16.0.0
+	 * @deprecated 18.0.0 Use IProviderManager::registerResourceProvider instead
```
Comment (ChristophWurst) [inline lib/public/Collaboration/Resources/IManager.php]: - [ ] Add to https://github.com/nextcloud/server/issues/17131 once merged

**Comment by ChristophWurst** [comment]
I like the idea. Small services are good services.

You'll get bonus points if you add tests :wink:

File: lib/private/Collaboration/Resources/Manager.php [hunk 4]
```diff
@@ -526,7 +507,8 @@ public function invalidateAccessCacheForProviderByUser(IProvider $provider, ?IUs
 	 * @param string $provider
 	 */
 	public function registerResourceProvider(string $provider): void {
-		$this->providers[] = $provider;
+		$this->logger->debug('Manager::registerResourceProvider is deprecated');
```
Comment (nickvergessen) [inline lib/private/Collaboration/Resources/Manager.php]: Please use the full class name and also log the `$provider` value so people can find out what this is about and which app fucked up

Comment (kesselb) [reply lib/private/Collaboration/Resources/Manager.php]: Yes

File: lib/private/Collaboration/Resources/ProviderManager.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+	public function __construct(ILogger $logger) {
+		$this->logger = $logger;
+	}
+
+	public function getResourceProviders(): array {
+		if ($this->providers !== []) {
+			foreach ($this->providers as $provider) {
+				try {
+					$this->providerInstances[] = \OC::$server->query($provider);
```
Comment (ChristophWurst) [inline lib/private/Collaboration/Resources/ProviderManager.php]: You can inject `\OCP\IServerContainer` via the constructor to make this mockable and testable

File: lib/private/Collaboration/Resources/ProviderManager.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+				} catch (QueryException $e) {
+					$this->logger->logException($e, [
+						'message' => 'Error when instantiating resource provider'
```
Comment (ChristophWurst) [inline lib/private/Collaboration/Resources/ProviderManager.php]:
```suggestion
						'message' => "Could not query resource provider $provider: " . $e->getMessage()
```

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: lib/private/Collaboration/Resources/ProviderManager.php (hunk 5 above)
Comment (kesselb) [reply lib/private/Collaboration/Resources/ProviderManager.php]: Yes. Need to mock all query calls then but we can run the tests without a proper di container.

File: tests/lib/Collaboration/Resources/ManagerTest.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+		$this->logger = $this->createMock(ILogger::class);
+		$this->providerManager = $this->createMock(IProviderManager::class);
+
+		/** @var IDBConnection $connection */
+		$connection = $this->createMock(IDBConnection::class);
+		$this->manager = new Manager($connection, $this->providerManager, $this->logger);
+	}
+
+	/**
+	 * @deprecated 18.0.0
```
Comment (ChristophWurst) [inline tests/lib/Collaboration/Resources/ManagerTest.php]: a deprecated test function? what? :)

File: tests/lib/Collaboration/Resources/ProviderManagerTest.php [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+			->with($this->equalTo('InvalidResourceProvider'))
+			->willThrowException(new QueryException('A meaningful error message'));
+
+		$this->logger->expects($this->once())
+			->method('logException');
+
+		$this->providerManager->registerResourceProvider('InvalidResourceProvider');
+		$resourceProviders = $this->providerManager->getResourceProviders();
+
+		$this->assertCount(0, $resourceProviders);
```
Comment (ChristophWurst) [inline tests/lib/Collaboration/Resources/ProviderManagerTest.php]: FYI there's also `assertEmpty`

**Review by ChristophWurst — APPROVED** [review]

File: tests/lib/Collaboration/Resources/ProviderManagerTest.php (hunk 8 above)
Comment (kesselb) [reply tests/lib/Collaboration/Resources/ProviderManagerTest.php]: Thanks

**Review by ChristophWurst — APPROVED** [review]

**Comment by kesselb** [comment]
Hello @rullzer @juliushaertl :ping_pong:

**Review by juliusknorr — APPROVED** [review]
Nice :+1:

**Comment by ChristophWurst** [comment]
CI seems happy except for then the DB dies. Let's get this in!

**Comment by ChristophWurst** [comment]
@kesselb mind squashing the commits?

**Comment by kesselb** [comment]
CI don't like me today ;) Could someone of you press merge?

File: tests/lib/Collaboration/Resources/ManagerTest.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+class ManagerTest extends TestCase {
+
+	/** @var ILogger */
+	protected $logger;
+	/** @var IProviderManager */
+	protected $providerManager;
+	/** @var IManager */
+	protected $manager;
+
+	public function setUp() {
```
Comment (nickvergessen) [inline tests/lib/Collaboration/Resources/ManagerTest.php]:
```suggestion
	protected function setUp(): void {
```

File: tests/lib/Collaboration/Resources/ProviderManagerTest.php [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+class ProviderManagerTest extends TestCase {
+
+	/** @var IServerContainer */
+	protected $serverContainer;
+	/** @var ILogger */
+	protected $logger;
+	/** @var IProviderManager */
+	protected $providerManager;
+
+	public function setUp() {
```
Comment (nickvergessen) [inline tests/lib/Collaboration/Resources/ProviderManagerTest.php]:
```suggestion
	protected function setUp(): void {
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]
Some small little fixes needed because we updated phpunit in the meantime

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**